### DISCUSSION TITLE
Populating UserLabel and CodeValue

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#1820] [Q&A] [answered]
Our client is preparing iModels from dgn files created in standard MicroStation (without using any PP application).  I wanted to find out (so I can advise them), which attributes/properties need to be set in MicroStation for the element so that these values end up in CodeValue and UserLabel in the iModel for that element.

**Comment by cammac1803** [comment, marked as answer]
If the V8 element has an ECInstance on it, the connector will use the DisplayLabel of that ECInstance as the UserLabel of the imodel02 Element. Otherwise, the UserLabel is set to the description given by the V8 element's handler.  The standard MicroStation Connector does not set CodeValues on elements by default.

**Reply by ovaisa** [reply]
Client is using Connect Edition, does this apply to CE too.  Is there anything in CE that can help in populating these attributes.

**Reply by pmconne** [reply]
If the user is trying to define user labels in MicroStation that will transfer to the elements in the iModel, one option is to use [Item Types](https://docs.bentley.com/LiveContent/web/MicroStation%20Help-v13/en/ItemTypes.html).
Define an Item Type with some property that will store the label for the element to which it is attached.
Set its "Use Item Name for Element" to true.
Set its "Item Name" property to the name of the property that will hold the label (or, if all elements with this Item attached should have the same label, just enter the name directly).
Attach Items of this type to elements.
Optionally, change the label property of each elements' attached Item to change the elements' labels.

After doing so, when you select an element with an Item attached, you should see the label from the Item displayed in the Explorer and Properties dialogs.

**Reply by ovaisa** [reply]
Thank you Paul, this seems like a better solution.  Will give it a try.

We are not using Explorer or Properties dialog, our plan is to use these fields to hold external keys to retrieve info from external databases with additional info on components and display in iTwin.  Using UserLabel or CodeValue seems most efficient - if you feel there is a more efficient way, I request your guidance.

**Reply by ovaisa** [reply]
Is there any documentation on how MicroStation bridge/synchonizer maps dgn schema to imodel ECSchema.  Our intent is to enrich MicroStation model so that we can surface/use this information in iTwin viewer.  We are trying to reverse engineer where MicroStation information is ending up in imodel.  Any documentation on this will be really helpful in modeling  our information.

**Reply by ovaisa** [reply]
Thank you Paul, I am able to get the required info in UserLabel using the method you explained above.  Much appreciate your detailed explanation, even as a non MS user, I was able to follow and get it working quite easily.
